package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,547 @@
+  SEND_FD_COUNTS_8,
+  SEND_FD_COUNTS_9,
+  SEND_FD_SUMMARY,
+  SEND_DONE,
+  SEND_ERROR
+} send_state_t;
+
+typedef struct {
+  char id_str[16];
+  u16 pid;
+  u16 fd_count;
+  char cmdline[COMMAND_LINE_MAX];
+} pid_entry_t;
+
+typedef struct {
+  char file_str[128];
+  u16 fd_count;
+} file_entry_t;
+
+MAKE_TABLE_WRAPPER(pid, pid_entry_t);
+MAKE_TABLE_WRAPPER(file, file_entry_t);
+
+typedef struct {
+  send_state_t send_state;
+  u8 error_code;
+  u32 total_fd_count;
+  table_t *pid_table;
+  table_t *file_table;
+  const char *top10_pid_counts[ITEM_COUNT];
+  const char *top10_file_counts[ITEM_COUNT];
+} resq_state_t;
+
+typedef struct id_to_count {
+  u16 count;
+  const char *id;
+} id_to_count_t;
+
+enum {
+  FD_STATE_ID_STR = 0,
+  FD_STATE_EXE_FILE_STR,
+  FD_STATE_FD_STR,
+  FD_STATE_DONE,
+  FD_STATE_COUNT,
+};
+
+static bool is_real_file_path(const char *path)
+{
+  if (strncmp(path, "pipe", 4) == 0 || strncmp(path, "socket", 6) == 0)
+    return false;
+  else
+    return true;
+}
+
+static bool process_fd_entry(resq_state_t *state, const char *file_str, const char *id_str)
+{
+  pid_entry_t *entry = pid_table_get(state->pid_table, id_str);
+  state->total_fd_count++;
+  if (entry == NULL) {
+    pid_entry_t new_entry = {0};
+    int s = snprintf(new_entry.id_str, sizeof(entry->id_str), "%s", id_str);
+    if (s < 0 || (size_t)s >= sizeof(entry->id_str)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_WARNING, "file descriptor query: failed to copy pid string");
+#endif
+      return false;
+    }
+    unsigned long pid = 0;
+    if (!strtoul_all(10, id_str, &pid)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_WARNING,
+                  "file descriptor query: failed to convert pid str to number: %s",
+                  strerror(errno));
+#endif
+      return false;
+    }
+    new_entry.pid = (u16)pid;
+    size_t file_path_len = (size_t)(15 + s);
+    char cmdline_file[file_path_len];
+    s = snprintf(&cmdline_file[0], file_path_len, "/proc/%s/cmdline", id_str);
+    if (s < 0 || (size_t)s >= file_path_len) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_WARNING,
+                  "file descriptor query: failed to cmdline file string : %s",
+                  cmdline_file);
+#endif
+      return false;
+    }
+    cmdline_file[file_path_len - 1] = '\0';
+    buffer_fn_t copy_cmdline = NESTED_FN(ssize_t, (char *buffer, size_t length, void *ctx), {
+      (void)ctx;
+      strncpy(&new_entry.cmdline[0], buffer, length);
+      for (unsigned int i = 0; i < length; i++) {
+        if (new_entry.cmdline[i] == '^' || new_entry.cmdline[i] == '@') {
+          new_entry.cmdline[i] = ' ';
+        }
+      }
+      return (ssize_t)length;
+    });
+
+    char buf[COMMAND_LINE_MAX] = {0};
+    run_command_t run_config = {.input = NULL,
+                                .context = NULL,
+                                .argv =
+                                  (const char *[]){"sudo", "/bin/cat", cmdline_file, "-v", NULL},
```

Comment:
This is adding 2 white spaces for the command line string (e.g. `"endpoint_router  --name  nmea  -f  /etc/endpoint_router/nmea_router.yml  "`) since the NULL escape sequence is `"^@"` (additionally, any ^ of @ that happens to be in the command line (while unlikely) will be translated to a ` `) -- I think we can replace `cat -v` with a small script wrapper to simplify this, e.g.:

```sh
#!/bin/ash
pid=$1; shift
cat /proc/$pid/cmdline | tr "\0" ' ' | sed 's@.$@@'
```